src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/CSharpConvertToRecordRefactoringProvider.cs

```diff
@@ -0,0 +1,623 @@
+        {
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, span, cancellationToken) = context;
+
+            var classDeclaration = await context.TryGetRelevantNodeAsync<TypeDeclarationSyntax>().ConfigureAwait(false);
+
+            if (classDeclaration == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            if (semanticModel.GetDeclaredSymbol(classDeclaration, cancellationToken) is not INamedTypeSymbol type ||
+                // if type is some enum, interface, delegate, etc we don't want to refactor
+                (type.TypeKind != TypeKind.Class && type.TypeKind != TypeKind.Struct) ||
+                // TODO: when adding other options it might make sense to convert a postional record to a non-positional one and vice versa
+                // but for now we don't convert records
+                type.IsRecord ||
+                // records can't be static and so if the class is static we probably shouldn't convert it
+                type.IsStatic ||
+                // make sure that there is at least one positional parameter we can introduce
+                !classDeclaration.Members.Any(m => ShouldConvertProperty(m, type)))
+            {
+                return;
+            }
+
+            var positionalTitle = FeaturesResources.Convert_to_positional_record;
+
+            var positional = CodeAction.Create(
+                positionalTitle,
+                cancellationToken => ConvertToParameterRecordAsync(document, type, classDeclaration, semanticModel, cancellationToken),
+                positionalTitle);
+            var nested = CodeAction.CodeActionWithNestedActions.Create(string.Format(FeaturesResources.Convert_0_to_record, type.Name), ImmutableArray.Create(positional), isInlinable: false);
+            context.RegisterRefactoring(nested);
+        }
+
+        private static async Task<Document> ConvertToParameterRecordAsync(
+            Document document,
+            INamedTypeSymbol originalType,
+            TypeDeclarationSyntax originalDeclarationNode,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            var codeGenerationService = document.GetRequiredLanguageService<ICodeGenerationService>();
+            var syntaxGenerator = document.GetRequiredLanguageService<SyntaxGenerator>();
+
+            var propertiesToMove = originalDeclarationNode.Members.Where(m => ShouldConvertProperty(m, originalType)).Cast<PropertyDeclarationSyntax>().AsImmutable();
+
+            var modifiedClassTrivia = GetModifiedClassTrivia(propertiesToMove, originalDeclarationNode);
+
+            var propertiesToAddAsParams = propertiesToMove.SelectAsArray(p =>
+                SyntaxFactory.Parameter(
```

Comment:
consider extracting the attribute logic out into a separate statement/function to help keep this clearer.